frame/nft-fractionalisation/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+	pub type AssetIdOf<T> =
+		<<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+	pub type AssetBalanceOf<T> =
+		<<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+	pub type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type Currency: Currency<Self::AccountId>;
+
+		/// Identifier for the collection of item.
+		type CollectionId: Member + Parameter + MaxEncodedLen + Copy + Incrementable;
+
+		/// The type used to identify a unique item within a collection.
+		type ItemId: Member + Parameter + MaxEncodedLen + Copy;
+
+		type AssetBalance: AtLeast32BitUnsigned
+			+ codec::FullCodec
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ sp_std::fmt::Debug
+			+ Default
+			+ From<u64>
+			+ IntegerSquareRoot
+			+ Zero
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ codec::HasCompact
+			+ From<u32>
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ PartialOrd
+			+ TypeInfo;
+
+		type Assets: Inspect<Self::AccountId, AssetId = Self::AssetId, Balance = Self::AssetBalance>
+			+ Create<Self::AccountId>
+			+ InspectEnumerable<Self::AccountId>
+			+ Mutate<Self::AccountId>
+			+ MutateMetadata<Self::AccountId>
+			+ Transfer<Self::AccountId>;
+
+		type Items: NonFungiblesInspect<
+				Self::AccountId,
+				ItemId = Self::ItemId,
+				CollectionId = Self::CollectionId,
+			> + NonFungiblesTransfer<Self::AccountId>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		#[pallet::constant]
+		type BuybackThreshold: Get<u32>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn assets_minted)]
+	// TODO: query amount minted from pallet assets instead of storing it locally.
+	// Add a public getter function to pallet assets.
+	pub type AssetsMinted<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, AssetBalanceOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn asset_to_nft)]
+	// TODO: store information about Asset ID and the corresponding Collection and Item ID.
+	pub type AssetToNft<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, (T::CollectionId, T::ItemId), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NFTLocked(T::CollectionId, T::ItemId),
+		NFTUnlocked(T::CollectionId, T::ItemId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetAlreadyRegistered,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(2).ref_time())]
+		/// Pallet's account must be funded before lock is possible!
+		/// 5EYCAe5gjC5dxKPbV2GPQUetETjFNSYZsSwSurVTTXidSLbh
```

Comment:
what is this account about? is that the pallet account ?

Reply:
Yes, exactly, if it's not funded originally, the lock fails. I am thinking to add a balance transfer of a small amount from the caller's account into the pallet's account in every lock.